src/HttpClientInterception/Matching/DelegatingMatcher.cs

```diff
@@ -30,5 +30,11 @@ internal DelegatingMatcher(Func<HttpRequestMessage, Task<bool>> predicate)
         /// <inheritdoc />
         public override async Task<bool> IsMatchAsync(HttpRequestMessage request)
             => await _predicate(request).ConfigureAwait(false);
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            return _predicate.GetHashCode();
+        }
```

Comment:
What's the specific use case that required this to be changed?

Is this something not covered by this?

https://github.com/justeat/httpclient-interception/blob/dc1fb4e9ee276856d21ac5f65615b5639b913674/src/HttpClientInterception/HttpClientInterceptorOptions.cs#L399-L404

Reply:
For `InternalMatcher` is instantiated new object every time that `ConfigureMatcher` is called. Two different matcher objects would be instantiated (here `DelegatingMatcher`) for Register and Deregister. So, it needs to using predictor instead of wrapper object of the predictor in order to get hash code and compare registered user mathcer.

In another word, in `BuildKey`, it needs to get an equal hash code of `InternalMatcher` in both Register and Deregister.

Reply:
OK - I'll have a think about this a bit further, as I'm not 100% sure that it's the _right_ fix.

The `NullReferenceException` is happening due to my assumption that there were no code paths where the internal matcher hadn't been set (which was wrong). I'm concerned that changing this behaviour will affect the referential integrity in other use cases and break something else.

Reply:
Thank you, I value your feedback, so let me know what you think!

Reply:
Sorry for taking so long with this issue, I haven't had the time to dedicate to get my head around the underlying issue and see if there's a different fix or a change needed to this one (or whether it's actually fine!).

I'll try and get around to it at some point this month.

Reply:
That is very gratifying, please let me know if I can be of any help.